[aoti x with_effect token] Unbacked symint and register lowering

yushangdi opened the pull request:
Differential Revision: D70022208


- When resolving unbacked symints in ExternKernel for with_effect, we need to ignore the first item in the binding path, because the `example_output` doesn't contain the effect token, but the binding paths do. 
- Similarly, `node.meta["val"]` contains the effect token, so when we compute_unbacked_bindings, we need to remove that effect token

- For `torch.ops.higher_order.with_effects`'s lowering, we should not extract the items out of an list (i.e. `*result` vs `result`). The `get_attr` nodes consider the result to be in the list format.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

pianpwk commented on torch/_inductor/ir.py:
```diff
@@ -5113,9 +5113,18 @@ def unflatten_args(new_tensor_args, new_non_tensor_args):  # type: ignore[no-unt
+                node_meta_val = node_meta_val[1]  # remove the first effect token
+            rebind_unbacked(
+                shape_env, V.current_node, example_output, remove_effect_token_in_path
+            )
```
Could we maybe modify V.current_node.meta["unbacked_bindings"] here instead? This feels like a very special case, I'd prefer if the flag didn't get passed into rebind_unbacked. Otherwise looks great!

yushangdi replied:
> Could we maybe modify V.current_node.meta["unbacked_bindings"] here instead? This feels like a very special case, I'd prefer if the flag didn't get passed into rebind_unbacked. Otherwise looks great!

@pianpwk Fixed now, thanks! (this Diff contains a lot of dependency changes, D70022208 might be easier to look at)

yushangdi committed:
[aoti x with_effect token] Unbacked symint and register lowering (#147656)

Summary:

- When resolving unbacked symints in ExternKernel for with_effect, we need to ignore the first item in the binding path, because the `example_output` doesn't contain the effect token, but the binding paths do.
- Similarly, `node.meta["val"]` contains the effect token, so when we compute_unbacked_bindings, we need to remove that effect token

- In `get_effect_key(op, args, kwargs)`, we should look at both args and kwargs.

Test Plan:
CI

```
buck run //caffe2/torch/fb/sparsenn:sigrid_test -- -r test_transform_torch_bind

buck run fbcode//mode/dev-nosan //caffe2/test:test_export  -- -r test_real_tensor_for_max_op_training_ir_to_decomp

buck run fbcode//mode/dev-nosan //caffe2/test/inductor:torchbind -- -r test_torchbind_aot_compile$
```


Tested together with D70013257

Reviewed By: angelayi

Differential Revision: D70022208

yushangdi commented:
@pytorchbot merge -i
